torchchat/utils/generator.py

```diff
@@ -0,0 +1,399 @@
+        )
+
+        # Validate that all image prompts exist before expensive model load
+        if image_prompts := getattr(args, "image_prompts", None):
+            non_existent_image_prompts = [
+                image_prompt
+                for image_prompt in image_prompts
+                if (not os.path.exists(image_prompt))
+            ]
+            if non_existent_image_prompts:
+                raise RuntimeError(
+                    f"Image prompt {non_existent_image_prompts} does not exist"
+                )
+
+        return cls(
+            prompt=getattr(args, "prompt", ""),
+            encoded_prompt=None,
+            image_prompts=image_prompts,
+            chat_mode=args.chat,
+            gui_mode=args.gui,
+            num_samples=getattr(args, "num_samples", 1),
+            max_new_tokens=args.max_new_tokens,
+            top_k=args.top_k,
+            temperature=args.temperature,
+            compile=args.compile,
+            compile_prefill=args.compile_prefill,
+            speculate_k=args.speculate_k,
+            sequential_prefill=sequential_prefill,
+            max_autotune=args.max_autotune,
+            is_torchtune_model=args.model and args.model.endswith("tune"),
+        )
+
+
+class Generator(object):
+    """
+    Base class for generators that can be used to generate text samples based on a pre-trained Transformer model and tokenizer.
+    """
+
+    def __init__(
+        self,
+        builder_args: BuilderArgs,
+        tokenizer_args: TokenizerArgs,
+        generator_args: GeneratorArgs,
+    ):
+        self.builder_args = builder_args
+        self.tokenizer_args = tokenizer_args
+        self.generate_args = generator_args
+
+        self.dtype = builder_args.precision
+
+        self.tokenizer = _initialize_tokenizer(self.tokenizer_args)
+
+        # Right now the assumption is only llama3 uses tiktokenizer and it
+        # must use tiktokenizer.
+        # Piggy backing off of this flag then for now to identify llama3
+        # without prompting user.
+        self.is_llama3_model = self.tokenizer_args.is_tiktoken
+        if self.is_llama3_model:
+            self.chat_formatter = Llama3ChatFormatter(self.tokenizer)
+            if generator_args.chat_mode:
+                logging.debug(
+                    "Llama3 model detected in chat mode. Using updated sentence schemas"
+                )
+        else:
+            self.chat_formatter = Llama2ChatFormatter(self.tokenizer)
+
+    @abstractmethod
+    def is_text_only(self) -> bool:
+        """
+        Returns True if the model is text-only, False otherwise.
+        """
+        raise NotImplementedError()
+
+    def _gen_model_input(
+        self,
+        prompt: Union[str | List[Any]],
+        image_prompts: Optional[List[str | Image.Image]] = None,
+        max_new_tokens: Optional[int] = None,
+        max_seq_len: Optional[int] = 2048,
+    ) -> Tuple[torch.Tensor, Optional[Dict[str, Any]]]:
```

Comment:
Something similar to generator was something we were looking into for Q4/H1, so I'm glad to see the intial work here lining up with my mental model.

Is `_gen_model_input` used in the `DistributedGenerator`? If not let's keep this in the `LocalGenerator` so that `Generator` is more succinct.

Reply:
We can be more aggressive with leaving things in LocalGenerator and Generator being thin, but that's something we can do later (but gen_model_input should be moved unless there's a need)

Reply:
Thanks for reviewing and feedback. The "copy and pastability" design aspect of TorchChat got to me through the recent discussions. Regarding _gen_model_input, I was planning to only move things into Generator that I could reuse in distributed. In my dev branch I actually pulled [chat into the base class](https://github.com/mreso/torchchat/blob/refactor/uplift_chat_into_base_generator_class/torchchat/utils/generator.py#L259
) to reuse it so _gen_model_input would need to live in Generator as well. I think in the end you could end up with only load_model, decode_one_token and decode_n_tokens and some model properties in the Local/DistributedGenerator which in some sense would make it "copy and pastability" again as everything lives in Generator.